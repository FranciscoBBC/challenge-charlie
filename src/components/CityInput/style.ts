import styled from 'styled-components';

export const InputHolder = styled.div<{ border: boolean }>`
    display: flex;
    flex: 1;
    background: #fff;
    border: ${({ border }) => (border ? '3px solid #ff0000' : 'none')};
    width: 100%;
    align-items: center;
    justify-content: flex-start;
    height: 50px;
    opacity: 85%;
    box-shadow: 0 1px 5px #888888;
    z-index: 1;
`;

export const IconWrapper = styled.div`
    padding: 0 5px 0 5px;
`;

export const InputWrapper = styled.div`
    width: 100%;
`;

export const CustomInput = styled.input`
    width: 90%;
    border: none;
    font-size: 26px;
    font-weight: bold;
    color: #8d8986;
    :focus {
        outline: none;
    }
    ::placeholder {
        color: rgb(183, 183, 183);
    }
`;
