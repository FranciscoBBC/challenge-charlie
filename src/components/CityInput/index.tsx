import React, { useMemo, useCallback, useState } from 'react';
import { BiCompass } from 'react-icons/bi';
import GeoLocationService from 'services/GeoLocationService';
import * as S from './style';

type CityInputProps = {
    state: string;
    city: string;
    // eslint-disable-next-line no-unused-vars
    onChangeCity: (location: any) => void;
};

// usado para fazer o controle do debounce
let timeoutControl: any;

const CityInput = ({
    state = 'Rio de Janeiro',
    city = 'Rio de Janeiro',
    onChangeCity
}: CityInputProps) => {
    const [inputWarning, setInputWarning] = useState(false);
    const cityPlaceHolder = useMemo(() => {
        if (state && city) return `${state}, ${city}`;
        return '';
    }, [city, state]);

    const handleChange = useCallback(
        async event => {
            if (timeoutControl) clearTimeout(timeoutControl);
            timeoutControl = setTimeout(async () => {
                const { value } = event.target;
                if (value.length > 0) {
                    try {
                        const response =
                            await GeoLocationService.getPositionByCityName(
                                value
                            );
                        if (response) {
                            setInputWarning(false);
                            onChangeCity(response);
                        } else {
                            setInputWarning(true);
                        }
                    } catch (error) {
                        setInputWarning(true);
                        console.error(error);
                    }
                }
            }, 1000);
        },
        [onChangeCity]
    );

    return (
        <S.InputHolder data-testid="city-input-holder" border={inputWarning}>
            <S.IconWrapper>
                <BiCompass color="8d8986" size="50px" />
            </S.IconWrapper>
            <S.InputWrapper>
                <S.CustomInput
                    type="text"
                    data-testid="city-input"
                    onChange={handleChange}
                    placeholder={cityPlaceHolder}
                />
            </S.InputWrapper>
        </S.InputHolder>
    );
};

export default CityInput;
