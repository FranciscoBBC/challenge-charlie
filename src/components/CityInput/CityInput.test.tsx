import React from 'react';
import { screen, render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import CityInput from '.';

describe('<CityInput />', () => {
    it('should render input', () => {
        const { container } = render(
            <CityInput
                state="São Paulo"
                city="São Paulo"
                onChangeCity={value => jest.fn(value)}
            />
        );

        expect(screen.getByTestId('city-input')).toBeTruthy();

        expect(screen.getByTestId('city-input')).toHaveAttribute(
            'placeholder',
            'São Paulo, São Paulo'
        );

        expect(container.firstChild).toMatchSnapshot();
    });

    it('should return a valid city', () => {
        const callbackFn = jest.fn(value => value.city);

        render(
            <CityInput
                state="São Paulo"
                city="São Paulo"
                onChangeCity={callbackFn}
            />
        );

        const input = screen.getByTestId('city-input');
        userEvent.type(input, 'Belo Horizonte');

        expect(input).toHaveValue('Belo Horizonte');

        setTimeout(() => {
            expect(callbackFn).toHaveBeenCalledTimes(1);
            expect(callbackFn).toHaveReturnedWith('Belo Horizonte');
        }, 3000);
    });

    it('should return a invalid value', () => {
        const callbackFn = jest.fn(value => value.city);

        render(
            <CityInput
                state="São Paulo"
                city="São Paulo"
                onChangeCity={callbackFn}
            />
        );

        const input = screen.getByTestId('city-input');
        userEvent.type(input, '1234567891');

        expect(input).toHaveValue('1234567891');

        setTimeout(() => {
            expect(callbackFn).toHaveBeenCalledTimes(1);
            expect(callbackFn).toHaveReturnedWith(null);
            expect(screen.getByTestId('city-input-holder')).toHaveStyle({
                border: '3px solid #ff0000'
            });
        }, 3000);
    });
});
