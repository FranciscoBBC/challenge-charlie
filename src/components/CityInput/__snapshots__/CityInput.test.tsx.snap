// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CityInput /> should render input 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  background: #fff;
  border: none;
  width: 100%;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  height: 50px;
  opacity: 85%;
  box-shadow: 0 1px 5px #888888;
  z-index: 1;
}

.c1 {
  padding: 0 5px 0 5px;
}

.c2 {
  width: 100%;
}

.c3 {
  width: 90%;
  border: none;
  font-size: 26px;
  font-weight: bold;
  color: #8d8986;
}

.c3:focus {
  outline: none;
}

.c3::-webkit-input-placeholder {
  color: rgb(183,183,183);
}

.c3::-moz-placeholder {
  color: rgb(183,183,183);
}

.c3:-ms-input-placeholder {
  color: rgb(183,183,183);
}

.c3::placeholder {
  color: rgb(183,183,183);
}

<div
  class="c0"
  data-testid="city-input-holder"
>
  <div
    class="c1"
  >
    <svg
      color="8d8986"
      fill="currentColor"
      height="50px"
      stroke="currentColor"
      stroke-width="0"
      viewBox="0 0 24 24"
      width="50px"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M12 2C6.486 2 2 6.486 2 12s4.486 10 10 10 10-4.486 10-10S17.514 2 12 2zm0 18c-4.411 0-8-3.589-8-8s3.589-8 8-8 8 3.589 8 8-3.589 8-8 8z"
      />
      <path
        d="m8 16 5.991-2L16 8l-6 2z"
      />
    </svg>
  </div>
  <div
    class="c2"
  >
    <input
      class="c3"
      data-testid="city-input"
      name="city-input"
      placeholder="São Paulo, São Paulo"
      type="text"
    />
  </div>
</div>
`;
